packages/contracts/contracts/src/Policy.sol

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.27;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IPolicy} from "./interfaces/IPolicy.sol";
+
+/// @title Policy
+/// @notice Implements a base policy contract that protects access to a target contract
+/// @dev Inherits from OpenZeppelin's Ownable and implements IPolicy interface
+///
+/// This contract serves as a base for implementing specific policy checks that must be
+/// satisfied before interacting with a protected target contract. It provides core
+/// functionality for managing the protected target address and access control.
+abstract contract Policy is IPolicy, Ownable(msg.sender) {
+    /// @notice The policy-protected contract address.
+    /// @dev This address can only be set once by the owner.
+    /// For example, the target is a Semaphore group that requires the subject
+    /// to meet certain criteria in order to join the group.
+    address internal target;
+
+    /// @notice Restricts function access to only the target contract.
+    /// @dev Throws TargetOnly error if called by any other address.
+    modifier onlyTarget() {
+        if (msg.sender != target) revert TargetOnly();
+        _;
+    }
+
+    /// @notice Sets the target contract address.
+    /// @dev Can only be called once by the owner.
+    /// @param _target Address of the contract to be protected by this policy.
+    /// @custom:throws ZeroAddress if _target is the zero address.
+    /// @custom:throws TargetAlreadySet if target has already been set.
+    /// @custom:emits TargetSet when target is successfully set.
+    function setTarget(address _target) public virtual onlyOwner {
```

Comment:
why this public vs external? saw that all other public facing functions are marked as external aside from this

Reply:
oh, just because `setTarget()` was supposed to be called somewhere in other smart contract logic. But since it must be `onlyOwner` this cannot happen, so marking it as `external` is best, thanks!

Reply:
done